include/espi.h

```diff
@@ -0,0 +1,446 @@
+	/** Maximum supported frequency */
+	enum espi_freq max_freq;
+	/** Supported channels */
+	enum espi_channel channel_caps;
+	/** Supported peripherals over channel 0 */
+	enum espi_peripheral periph_caps;
+};
+
+struct espi_callback;
+
+/**
+ * @typedef espi_callback_handler_t
+ * @brief Define the application callback handler function signature
+ *
+ * @param "struct device *dev" Device struct for the ESPI device.
+ * @param "struct espi_callback *cb" Original struct espi_callback
+ *        owning this handler
+ * @param "espi_event espi_evt" event details that trigger the callback handler
+ *
+ */
+typedef void (*espi_callback_handler_t)(struct device *dev,
+					struct espi_callback *cb,
+					struct espi_event espi_evt);
+
+
+/**
+ * @cond INTERNAL_HIDDEN
+ *
+ * Used to register a callback in the driver instance callback list.
+ * As many callbacks as needed can be added as long as each of them
+ * are unique pointers of struct espi_callback.
+ * Beware such structure should not be allocated on stack.
+ *
+ * Note: To help setting it, see espi_init_callback() below
+ */
+struct espi_callback {
+	/** This is meant to be used in the driver only
+	 */
+	sys_snode_t node;
+
+	/** Actual callback function being called when relevant. */
+	espi_callback_handler_t handler;
+
+	/** An event which user is interested in, if 0 the callback
+	 * will never be called. Such evt_mask can be modified whenever
+	 * necessary by the owner, and thus will affect the handler being
+	 * called or not.
+	 */
+	struct espi_event evt;
+};
+
+/**
+ * @cond INTERNAL_HIDDEN
+ *
+ * ESPI driver API definition and system call entry points
+ *
+ * (Internal use only.)
+ */
+typedef int (*espi_api_config)(struct device *dev, struct espi_cfg *cfg);
+typedef int (*espi_api_set_channel_status)(struct device *dev,
+					enum espi_channel ch);
+typedef bool (*espi_api_get_channel_status)(struct device *dev,
+					enum espi_channel ch);
+typedef int (*espi_api_set_vwire)(struct device *dev, enum espi_vw_signal vw,
+					u8_t level);
+typedef int (*espi_api_get_vwire)(struct device *dev, enum espi_vw_signal vw,
+					u8_t *level);
+
+typedef int (*espi_api_manage_callback)(struct device *dev,
+					struct espi_callback *callback,
+					bool set);
+typedef int (*espi_api_enable_callback)(struct device *dev,
+					struct espi_event *cfg);
+typedef int (*espi_api_disable_callback)(struct device *dev,
+					struct espi_event *cfg);
+typedef u32_t (*espi_api_get_pending_int)(struct device *dev);
+
+struct espi_driver_api {
+	espi_api_config config;
+	espi_api_get_channel_status get_channel_status;
+	espi_api_set_channel_status set_channel_status;
+	espi_api_set_vwire set_vwire;
+	espi_api_get_vwire get_vwire;
+	espi_api_manage_callback manage_callback;
+	espi_api_enable_callback enable_callback;
+	espi_api_disable_callback disable_callback;
+	espi_api_get_pending_int get_pending_int;
+};
+
+__syscall int espi_config(struct device *dev, struct espi_cfg *cfg);
```

Comment:
Need documentation comments for all APIs

Reply:
Also, are all these APIs supposed to be `INTERNAL_HIDDEN`? I would assume that applications are using these.

Reply:
Fixed